src/unix/pipe.c

```diff
@@ -379,3 +379,51 @@ int uv_pipe_chmod(uv_pipe_t* handle, int mode) {
 
   return r != -1 ? 0 : UV__ERR(errno);
 }
+
+
+int uv_pipe(uv_os_fd_t fds[2], int read_flags, int write_flags) {
+  int err;
+#if defined(__FreeBSD__) || defined(__linux__)
+  int flags = O_CLOEXEC;
+
+  if ((read_flags & UV_NONBLOCK_PIPE) && (write_flags & UV_NONBLOCK_PIPE))
+    flags |= UV_FS_O_NONBLOCK;
+
+  if (pipe2(fds, flags))
+    return UV__ERR(errno);
+
+  if (flags & UV_FS_O_NONBLOCK)
+    return 0;
+#else
+  if (pipe(fds))
+    return UV__ERR(errno);
+
+  if ((err = uv__cloexec(fds[0], 1)))
+    goto fail;
+
+  if ((err = uv__cloexec(fds[1], 1)))
+    goto fail;
+#endif
+
+  if (read_flags & UV_NONBLOCK_PIPE)
+    if ((err = uv__nonblock(fds[0], 1)))
+      goto fail;
+
+  if (write_flags & UV_NONBLOCK_PIPE)
+    if ((err = uv__nonblock(fds[1], 1)))
+      goto fail;
+
+  return 0;
+
+fail:
+  uv__close(fds[0]);
+  uv__close(fds[1]);
+  return err;
```

Comment:
It'd be better to not clobber `fds` on error. Same suggestion applies to `uv_socketpair()`.